_posts/2017-04-20-how-software-works-part1.md

```diff
@@ -0,0 +1,196 @@
+of software: Machine code.
+
+### 1 10 20 0
+
+From this point on, we will assume that a device exists which can run machine code.
+It has a CPU, RAM memory, buses to transfer data etc. Generally what one would call a computer
+architecture. For example, I'm writing this blog on a MacBook Pro which has one of the most common
+computer architectures called x86. The architecture has intrinsic, hardware-level routines to
+do very simple computations, move data between RAM and the CPU, manipulate registers etc. We will
+get to what all of that means but for now let's just assume that the hardware is there.
+
+The architecture must provide a mechanism to run programs. From the time computers were invented to
+the 70s the main format to store programs was punchcards. Since physically punching cards is not
+the most convenient way to write programs, a better way had to be invented. Nowadays every program that runs directly
+on the hardware is stored as a binary file, literally as ones and zeroes. These files contain what is called
+machine code because they are sent directly to the machine to be executed.
+
+This is where learning about binary pays off. While the machine code is stored as ones and zeroes,
+the CPU reads it one byte (8 bits) at a time. A sequence of bytes form what is called an instruction. These instructions
+conform to a certain format that the CPU is wired to read and execute. For example, let's make up an instruction
+set with the following format:
+
+```
+1 byte | 1 byte | 1 byte
+-------+--------+-------
+OPCODE | NUMBER | NUMBER
+```
+
+The word opcode refers to what type of operation we want the CPU to execute on the given arguments. Now let's define
+what our imaginary architecture can do and what opcodes one would use to invoke those routines.
+
+```
+Opcode | Instruction name
+-------+-----------------
+0      | HALT
+1      | ADD
+2      | SUB
+3      | MUL
+4      | DIV
+```
+
+So, our computer can add, substract, multiply and divide two numbers. There's also a code for halting the execution
+for when the program has finished. A simple program for this machine would look like this:
+
+`00000001 00001010 00010100 00000000`
+
+To make this more human-readable, let's convert the binary digits back to decimal:
+
+`1 10 20 0`
+
+When we execute this machine code the computer reads in the first byte and sees that it needs to execute the
+`ADD` instruction, taking arguments `10` and `20`. Once that is done, it sees that it has to `HALT` the execution.
+The CPU halts, concluding our first program on our imaginary computer.
+
+I hope this section showed you how computers read _binary_ programs in a certain format and execute the
+specified instructions on _binary_ data. Now, in reality no-one really writes machine code by hand. It is
+just way too arduous to type out numbers with no mneumonics associated with them. This is why assembly was invented.
```

Comment:
This section showed you how computers read _binary_ programs in a certain format and execute the specified instructions on _binary_ data. Now a days no one really writes machine code by hand.